core/src/message/init.rs

```diff
@@ -133,7 +133,7 @@ impl InitPacket {
     /// Create new InitPacket without gas.
     pub fn new(code_id: CodeId, salt: Salt, payload: Payload, value: Value) -> Self {
         Self {
-            program_id: ProgramId::generate(code_id, salt.inner()),
+            program_id: ProgramId::generate(0u32, code_id, salt.inner()),
```

Comment:
Nope.

`InitPacket::new` must accet nonce, that will be used for program id generation

Reply:
The issue is that when we take in a InitPacket, like in the core, it doesn't have a message id yet. What it does, is it gives over the InitPacket to the create_program function, which means the ProgramId would already be created, the ext create_program actually makes the messageid, which means that I'd need to change around the way InitPacket works and is used. Should I do that?

Reply:
So `InitPacket::new` should not take in the nonce, instead, it will keep the CodeId and salt, and not generate a new ProgramId. Then, in the Ext `create_program` function, the data from the packet is used with the newly made message id to make the ProgramId.

Reply:
> The issue is that when we take in a InitPacket, like in the core, it doesn't have a message id yet. What it does, is it gives over the InitPacket to the create_program function, which means the ProgramId would already be created, the ext create_program actually makes the messageid, which means that I'd need to change around the way InitPacket works and is used. Should I do that?

I guess you should create a new program id for `gr_create_program` from currently executing message id.
I think there's a way to obtain it from ext in `core-backend` sys-call implementation.

Reply:
Or you can play around `InitPacket` type somehow. For example, `Externalities::create_program` can accept not the `InitPacket`, but something like `SemiInitPacket` (choose name by yourself), which later must be converted to init packet using the nonce/message id to generate program id.

You aren't required to implement exactly that, but use similar approach with types.

Reply:
> So `InitPacket::new` should not take in the nonce, instead, it will keep the CodeId and salt, and not generate a new ProgramId. Then, in the Ext `create_program` function, the data from the packet is used with the newly made message id to make the ProgramId.

Yeah, nice idea. I described something similar in the previous comment.